Add option to use pod name rather than IP address for Kubernetes

bakins opened the pull request:
This PR is an idea in code form. 

### 1. What does this pull request do?

This adds an option to use the name of a pod as the hostname rather the hostname for lookups when pods mode is set to insecure or verified for the Kubernetes plugin.

### 2. Which issues (if any) are related?

https://github.com/kubernetes/kubernetes/issues/47992 - I did not find a related coredns issue doing a quick issue search.

We are also using airflow and this change enables DNS for pod FQDNs without setting`hostname` in the pod specs. As we use deployments, we do not want each pod to have the same hostname.

### 3. Which documentation changes (if any) need to be made?

We would need to document the additional, optional, argument to `pods` in the Kubernetes plugin.

I was unsure if this should be a separate option or part of the `pods` mode.

As I said, this is to capture an idea in code form. It is fairly simple and has tests that pass.

miekg commented:
Reading through that whole thread, this seems a reasonable request.

chrisohaver commented:
This seems like a reasonable option.  This follows the endpoint naming spec, since the pod name is unique in the namespace.

To clarify, this is  "Adding option to use pod name rather than IP address for _k8s endpoint fqdns_" ...
It's changing the way we name endpoints.  The default behavior is to use the pod hostname, the optional behavior being added here is to use the pod name.  I think "endpointNameMode" might be a slightly better option name.

For example: `my-pod.my-service.namespace.svc.cluster.local` instead of the default `1-2-3-4.my-service.namespace.svc.cluster.local`

This does not change the way pod lookups function.
For example: `1-2-3-4.namespace.pod.cluster.local`

miekg commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -276,10 +277,13 @@ func (k *Kubernetes) Records(state request.Request, exact bool) ([]msg.Service,
 	if addr.Hostname != "" {
 		return strings.ToLower(addr.Hostname)
 	}
+	if podNameMode == "name" && addr.TargetRef != nil && addr.TargetRef.Name != "" {
```
Can you make "name" a constant, similar to add podMode ones

miekg commented on plugin/kubernetes/setup.go:
```diff
@@ -106,13 +106,24 @@ func kubernetesParse(c *caddy.Controller) (*Kubernetes, dnsControlOpts, error) {
 			case "pods":
 				args := c.RemainingArgs()
-				if len(args) == 1 {
+				k8s.podNameMode = "default"
```
default is just not setting this? I.e. no need for "default"?

johnbelamaric commented:
This changes the endpoint records; ie, with svc not pod type in the name. The pods config option is for the pod records. Maybe we need an endpoints config option instead.

bakins commented:
@chrisohaver this does allow lookups to work for queries such as `pod-phdc8.my-service.svc.cluster.local` if the pod is n th

bakins commented:
Changes made. I added an `endpoints` config option, as I agree that reflects its actual usage better.
I added documentation - I'd appreciate feedback on it.

I can squash the commits if/when ready to merge.

johnbelamaric commented on plugin/kubernetes/README.md:
```diff
@@ -65,6 +66,16 @@ kubernetes [ZONES...] {
+
+   * `name`: Return an A record if there exists a pod that is a valid endpoint
+    for the service.  This only works if `pods` is set to either `insecure` or
+    `verified`.  If `hostname` is set on a pod, then it always has precedence.
```
This should work regardless of the pods setting, I believe. We always watch endpoints, and we respond to their hostname records regardless of whether `pod` types are enabled.

We could add an option to `pods verified` that enables `pod-name.namespace.pod.cluster.local`.

johnbelamaric replied:
But that would be a separate PR...

bakins replied:
@johnbelamaric you are correct, it works if `pods` is disabled. Updating doc.

johnbelamaric commented on plugin/kubernetes/README.md:
```diff
@@ -65,6 +66,16 @@ kubernetes [ZONES...] {
 
+* `endpoints` **ENDPOINT-MODE** sets the mode for handling service endpoint A
+   records, e.g.
+   `pod-name.my-service.svc.cluster.local. in A 1.2.3.4` if the pod is an
```
Normally if the endpoint has a hostname then these will resolve. This change just makes them use the pod name if it exists and the hostname doesn't. I think that needs to be made clearer here.

bakins replied:
"If `hostname` is set on a pod, then it always has precedence." - is that enough?

miekg commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -276,10 +279,13 @@ func (k *Kubernetes) Records(state request.Request, exact bool) ([]msg.Service,
 	if addr.Hostname != "" {
 		return strings.ToLower(addr.Hostname)
 	}
+	if endpointNameMode == endpointNameModeName && addr.TargetRef != nil && addr.TargetRef.Name != "" {
```
Actually now that I think some more; endpointNameMode is just a boolean, no need to carry a string with us all time and to a string compare - this is either on or off.

miekg commented on plugin/kubernetes/setup_test.go:
```diff
@@ -32,6 +33,7 @@ func TestKubernetesParse(t *testing.T) {
 			defaultResyncPeriod,
 			"",
 			podModeDisabled,
+			"",
```
Please create a new parse test for just this option; this adds a whole bunch of empty strings for no purpose at all in this test.

miekg commented on plugin/kubernetes/setup.go:
```diff
@@ -104,6 +104,17 @@ func kubernetesParse(c *caddy.Controller) (*Kubernetes, dnsControlOpts, error) {
 		for c.NextBlock() {
 			switch c.Val() {
+			case "endpoints":
+				args := c.RemainingArgs()
```
I would reverse this and not use a switch:
~~~
if len(args) != 1 { 
     return nil, opts. c.ArgErr()
}
if args[0] != endpointNameModeName {
    return nil, opts. fmt.Errorf("wrong value for endpoints: %s")
}

bakins commented:
Changes made. Writing docs is the hardest part of any project for me, so let me know what you think.

chrisohaver commented on plugin/kubernetes/README.md:
```diff
@@ -65,6 +66,18 @@ kubernetes [ZONES...] {
      option requires substantially more memory than in insecure mode, since it will maintain a watch
      on all pods.
 
+* `endpoints` **ENDPOINT-MODE** sets the mode for handling service endpoint A
```
How about this instead...

* `endpoints` **ENDPOINT-MODE** sets the mode for naming endpoint A
   records, e.g. `endpoint-name.my-service.namespace.svc.cluster.local. in A 1.2.3.4`
   By default, the `endpoint-name` is set to the `hostname` of the endpoint, or if `hostname` is not set, the dashed form of the endpoint ip address (e.g. `1-2-3-4.my-service.namespace.svc.cluster.local.`) 

   Valid values for **ENDPOINT-MODE**:

   * `name`: If the endpoint hostname is not set, use the pod name of the pod targeted by the endpoint.  If there is no pod targeted by the endpoint, use the dashed ip address form.

johnbelamaric replied:
ok, but as @miekg said, since there is only one value for endpoints-mode, maybe we should just say that the whole directive is `endpoints pod names` or similar. That will simplify the config and docs a bit. I don't want to just say `endpoints` because it's not clear enough how it changes the behavior.

Questions:
 * Will it still respond to the dashed-ip name even if this is set? That is, is this an additional name or a replacement name? Should we pick one as the A record and have a CNAME for the other?
 * Which name shows up in the SRV record for the service?

chrisohaver replied:
Per my understanding of the code...

> Is this an additional name or a replacement name? 
  This is a replacement name, not an additional name.

> Which name shows up in the SRV record for the service?

`SRV` targets will point to the `A` records of the endpoints.

chrisohaver replied:
Agreed ... here is a "boolean" directive version...

* `endpoint_pod_names` Use the pod name of the pod targeted by the endpoint as the endpoint name in `A`  records, e.g. `endpoint-name.my-service.namespace.svc.cluster.local. in A 1.2.3.4`
   By default, the `endpoint-name` name selection is as follows: Use the `hostname` of the endpoint, or if `hostname` is not set, use the dashed form of the endpoint ip address (e.g. `1-2-3-4.my-service.namespace.svc.cluster.local.`) 
   If this directive is included, then name selection for endpoints changes as follows: Use the `hostname` of the endpoint, or if `hostname` is not set, _use the pod name of the pod targeted by the endpoint_.  If there is no pod targeted by the endpoint, use the dashed ip address form.

johnbelamaric replied:
👍 now @bakins please update the code to match :)

chrisohaver commented:
/integration (kicking off k8s integration tests)

miekg commented:
think we fleshed out what needs to be done here. Shall we wait with 0.9.10 so this can go in?

johnbelamaric commented:
No, let's just get 0.9.10 out ASAP.

bakins commented:
Changes made. Let me know if this all looks good and I'll squash.

johnbelamaric commented:
/integration

chrisohaver commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -70,6 +71,8 @@ const (
 	// DNSSchemaVersion is the schema version: https://github.com/kubernetes/dns/blob/master/docs/specification.md
 	DNSSchemaVersion = "1.0.1"
+	// endpointNameModeName is where we us the pod name rather than the IP for the endpoint hostname.
+	endpointNameModeName = "name"
```
We can remove `endpointNameModeName` now...

chrisohaver replied:
We can remove `endpointNameModeName` now...

bakins commented:
Okay, pushed hopefully the last change - removed endpointNameModeName. Let me know and I will squash, if that's what you want.

miekg commented:
We squash on merge, but feel free to do it yourself and force push the
branch.

On Nov 3, 2017 21:26, "Brian Akins" <notifications@github.com> wrote:

> Okay, pushed hopefully the last change - removed endpointNameModeName. Let
> me know and I will squash, if that's what you want.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coredns/coredns/pull/1190#issuecomment-341831447>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAVkW4fB6OgaGHuX0JC9wQYiIWG-X6yhks5sy4UCgaJpZM4QPn1r>
> .
>

bakins commented:
squashed and pushed.

miekg commented:
@johnbelamaric are you happy w/ this?

tomplus commented:
When are you going to release next version with this feature ? I'm looking forward to use it. Thanks.

johnbelamaric commented:
Yes, 1.0.0 will be released this week or early next week.

tomplus commented:
Thank you, it works like a charm !
